### ISSUE TITLE
bug: input focus scroll does not work when viewport resizing is disabled

### ISSUE THREAD

**Issue by Spinnenzunge** [ionic-team/ionic-framework#18532]
# Bug Report
If KeyboardResize is set to false the keyboard overlaps the webview and does not resize it. The Content scrolling on input focus does not work in this case.
As it is at the moment, I prefer the keyboard overlapping the webview and not resizing it because resizing feels clunky with jumped animations, tab-bar hiding bugs, etc.

**Ionic version:**

[x] **4.5.0**

**Current behavior:**
Content scrolling not available with KeyboardResize=false because webview is full-size

**Expected behavior:**
If KeyboardResize is set to false, ion-content adds a keyboard-offset as padding-bottom so the content is bigger and content scrolling works.

**Steps to reproduce:**
Add `<preference name="KeyboardResize" value="false" />` to config.xml, create a page with a long form that uses more than half of the screen

**Related code:**
Config.xml
```
<preference name="KeyboardResize" value="false" />
```

Page.html

```
<ion-content>
  <ion-list>
    <ion-item>
      <ion-label position="stacked">
        <h3>Strasse, Nr.</h3>
      </ion-label>
      <ion-input type="text" name="street" inputmode="text" autocomplete="street-address" [(ngModel)]="address.street"></ion-input>
    </ion-item>

    <ion-item>
      <ion-label position="stacked">
        <h3>PLZ</h3>
      </ion-label>
      <ion-input type="number" name="postalCode" pattern="[0-9]*" autocomplete="postal-code"
                 [(ngModel)]="address.postalCode"></ion-input>
    </ion-item>

    <ion-item>
      <ion-label position="stacked">
        <h3>Ort</h3>
      </ion-label>
      <ion-input type="text" name="location" inputmode="text" autocomplete="locality" [(ngModel)]="address.location"></ion-input>
    </ion-item>

    <ion-item>
      <ion-label position="stacked">
        <h3>Land</h3>
      </ion-label>
      <ion-input type="text" name="country" inputmode="text" autocomplete="country" [(ngModel)]="address.country"></ion-input>
    </ion-item>
  </ion-list>
</ion-content>
```

**Other information:**

![Image from iOS](https://user-images.githubusercontent.com/11189008/59420249-d9e4b500-8dcc-11e9-9581-7cb40cdb130d.png)

**Ionic info:**

```
Ionic:

   ionic (Ionic CLI)             : 4.12.0 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.5.0
   @angular-devkit/build-angular : 0.12.4
   @angular-devkit/schematics    : 7.2.4
   @angular/cli                  : 7.2.4
   @ionic/angular-toolkit        : 1.4.1

Cordova:

   cordova (Cordova CLI) : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms     : android 8.0.0, ios 5.0.1
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 2.4.1, (and 14 other plugins)

System:

   Android SDK Tools : 26.1.1 (/Users/pascalgraf/Library/Android/sdk)
   ios-deploy        : 1.9.4
   ios-sim           : 8.0.1
   NodeJS            : v10.14.1 (/usr/local/bin/node)
   npm               : 6.9.0
   OS                : macOS Mojave
   Xcode             : Xcode 10.2.1 Build version 10E1001

```

**Comment by ArkeshGKalathiya** [comment]
Did you find any solution?, I am having same issue. Seems like it's not adding --keyboard-offset.

**Comment by Spinnenzunge** [comment]
As I see it, --keyboard-offset does not even exist if KeyboardResize is set to false, as the Keyboard overlays the App.

**Comment by danbn-100295** [comment]
Same problem,any thing new for fix it ?

**Comment by uncvrd** [comment]
I am having a similar issue, if I use a native input on my page, the viewport does not adjust to have the input on screen. However, if I use `IonInput` **it does** animate the viewport to show the input instead of it being hidden behind the keyboard. I tried digging in to the component source code to see what could be fixing that, but no luck yet.

**Comment by danbn-100295** [comment]
Change @ionic/angular to this version work for me
`"@ionic/angular": "5.2.2",`

**Comment by liamdebeasi** [comment]
Can everyone try the following dev build and let me know if it resolves the issue?

**Angular**
```
npm install @ionic/angular@5.5.0-dev.202011171750.33768e1
```

**React**
```
npm install @ionic/react@5.5.0-dev.202011171750.33768e1 @ionic/react-router@5.5.0-dev.202011171750.33768e1
```

**Vue**
```
npm install @ionic/vue@5.5.0-dev.202011171750.33768e1 @ionic/vue-router@5.5.0-dev.202011171750.33768e1
```

**Stencil/Vanilla JavaScript**
```
npm install @ionic/core@5.5.0-dev.202011171750.33768e1
```

**Comment by Spinnenzunge** [comment]
I installed the dev build for an Ionic Angular project (https://github.com/Spinnenzunge/ionic-framework-issue-18532) and it seems to work with disabled viewport resizing with this build. Can't say anything for vanilla, react and vue. Having an active cursor and scrolling the webview still feels a little bit janky but I guess that does not have anything to do with Ionic.

**Comment by liamdebeasi** [comment]
Thanks for testing! The input focus scroll should not impact Android since I think it always resizes the webview (I don't think there's a way to turn that off easily besides `android:windowSoftInputMode="adjustResize"`?)

**Comment by Spinnenzunge** [comment]
Exactly, you can't do it with the cordova keyboard plugin preference but you can add it directly to the AndroidManifest via the edit-config tag in config.xml and set the windowSoftInputMode like that. We used to have that in our App, so I feel like it is worth checking the Input Focus Scrolling as well on Android with the keyboard resizing disabled.

**Comment by liamdebeasi** [comment]
Makes sense. I tested your repo, and it seems to work fine on Android. Would love to have other people test too!

For anyone else testing: The scroll assist is only enabled by default on iOS, so if you want to use it on Android you will need to add `inputShims: true` to your config:

```js
IonicModule.forRoot({ inputShims: true })
```

**Comment by aparajita** [comment]
@liamdebeasi While the dev version does with the delay I noticed, in some ways it's worse. Try it with https://github.com/aparajita/ionic-input-problem.

- Inputs 1-5 are **always** scrolling the input to the top. Is that what you meant to do?
- On the main page, input 6 scrolls down to be just above the keyboard, where it should be. But it's rather disconcerting for the user to jump from the top in the other inputs to the bottom.
- In the modal, input 6 is cut off by the accessory bar.

The whole UX is rather jumpy now. I'm sorry to say I wouldn't want to present this to my users. I have a solution using non-scrollable ion-content and tracking focus + keyboard show that is working quite well, I'll have to stick with that for now. I'll pass it along when I have it fully tested, it may work well with scrollable ion-content as well.

**Comment by liamdebeasi** [comment]
Can you provide some steps to reproduce? I am not seeing that on my end (maybe I need to test with an iPhone 7 here too?)

**Comment by aparajita** [comment]
@liamdebeasi Try it in an iPhone 8 simulator, I get the same result there as on my iPhone 7.

**Comment by aparajita** [comment]
@liamdebeasi To reproduce it, run my demo app, and tap sequentially in the inputs from top to bottom.

**Comment by aparajita** [comment]
Hey @liamdebeasi, my solution to the forcing inputs into view is here: https://github.com/aparajita/ws-input-visibility

Unfortunately you won't be able to see how it acts unless you disable what you're doing. A video of how it looks in action using ionic/vue@0.0.9 is [here](https://drive.google.com/file/d/1LesJkfT2CSDDx9wj8PB79taaO4hkEwtf/view?usp=sharing).

To use it, do this in App.vue:

```ts
import { ensureVisibility } from 'ws-input-visibility';
...
setup() {
   onBeforeMount(() => {
      ensureVisibility();
  }
}
```

That's all that's necessary. Here's why I prefer this solution:

- It works with both scrollable and non-scrollable ion-content.
- It moves the input the minimum distance possible to place it in view.
- If an input is already completely in view, it is not moved.
- Priority is given first to placing the input completely above the keyboard. If it already is above the keyboard, or moving it would cause the top of the input to be out of view, the top is moved into view.
- It works identically with clicking and tab motion.
- The transitions can be configured individually for show/hide and ios/android, or removed entirely.
- The margin between the keyboard and the input is configurable.

I hope you will consider using this code. From a UX perspective it is a more predictable experience. 🙏

**Comment by liamdebeasi** [comment]
Thanks, I appreciate you digging in and helping out with fixing this issue. I agree that your solution leads to a smoother transition, but what concerns me with this approach is that it is using `transform` which is going to physically move `ion-content` offscreen. What this means is that if someone has the keyboard open in a form, they will be unable to scroll to the top-most input since the content was partially transformed offscreen.

Scroll Assist was written several years ago (around the Ionic 2/3 days 🙂 ) so it's probably time for a revamp in how we handle it. I will do some research as to how we can better implement Scroll Assist in Ionic Framework so it's less janky.

**Comment by liamdebeasi** [comment]
Looks like Apple has a guide on this for apps built with Swift/Objective-C, so I think this is where I will start my research: https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/KeyboardManagement/KeyboardManagement.html

**Comment by aparajita** [comment]
Thanks for taking a look @liamdebeasi, I really want to get this solved.

I only used `transform` as a temporary hack because I couldn't get scrolling to work — I tried calling `ionContent.scrollByPoint()` without success. The exact same code will work with a scroll by replacing the lines in `moveElementIntoView()` that set the `transform` and `transition`. Everything else can remain the same, it's the calculations that matter.

I just read the Apple docs you referenced. They say it's up to us to scroll the input into view, there is no real guidance other than that.

If you tell me where the Scroll Assist code is I'll look into making a PR that integrates my code.

**Comment by liamdebeasi** [comment]
I spent some time reworking Scroll Assist, and I fixed a few things that should significantly improve the user experience:

1. We now listen on either the input or its parent `ion-item` (if one exists) to initiate scroll assist. This resolved an issue where scroll assist would sometimes not work because users tapped `ion-item`. `ion-item` delegates focus to components like `ion-input`, so the keyboard would have still opened.

2. Updated scroll padding so that it works with `ion-input`. This is the fix that I mentioned in https://github.com/ionic-team/ionic-framework/issues/18532#issuecomment-729223021. This should resolve the thread's issue where scroll assist was not working when viewport resizing was disabled.

3. We no longer wait for the keyboard to finish opening or for the webview to resize. We can do this because by the time we need to scroll the input into view, the scroll padding utility has already added the scroll height we need. This should improve the performance of scroll assist and make it seem more responsive.

4. We have improved our methods for determining how much we need to scroll the input by. If you are in Capacitor, we use the precise keyboard height given in the `keyboardWillOpen` event. If you are in a browser/PWA we use the approximate keyboard height given in the `keyboardHeight` config option. This should improve the reliability of the scrolling and it should resolve the issue where tapping inputs would always cause the content to scroll upwards.

Due to the significance of scroll assist, I have decided that these changes warrant more testing than usual. As a result, I am going to remove this issue from its existing milestone. My plan is to introduce this change as an experimental, opt-in feature in an upcoming feature release of Ionic Framework with the stable version shipping in Ionic Framework v6.

I am doing some additional testing, but I will provide a dev build here soon.

**Comment by aparajita** [comment]
Thanks @liamdebeasi for taking up the challenge. Looking forward to testing it out.

**Comment by liamdebeasi** [comment]
Ok here is my first draft of the revised scroll assist. A few things to note about this dev build:

1. I am currently focusing on getting this working with webview resizing disabled. As a result, scroll assist will not work if webview resizing is enabled currently. I have plans to address this later. Here is how to disable webview resizing:

**Capacitor**
**capacitor.config.json**
```json
{
  "plugins": {
    "Keyboard": {
      "resize": "none"
    }
  }
}
```

**Cordova**
**config.xml**
```xml
<preference name="KeyboardResize" value="false" />
````

2. When running this in a web browser or as a PWA, we do not know the exact height of the keyboard and need to guess. As a result, the amount we scroll by will be slightly more than needed. At the moment I do not have plans to address this since there is no fast way to get the exact keyboard height. I can use Visual Viewport, but that would mean not being able to scroll until the keyboard is fully open causing scroll assist to only start after a ~400ms delay.

3. When running the demo that @aparajita provided in a web browser, focusing input 6 then input 5 causes Safari to shift the entire view. I am still investigating why this is happening. This does not impact Capacitor/Cordova. Additionally, a similar behavior appears when using the older scroll assist.

4. I have not addressed the following issue from https://github.com/ionic-team/ionic-framework/issues/22513 yet:

> On both Chrome and iOS, the content of an ion-textarea with left padding shifts left and then right when focused.


Here is the dev build for those who would like to test: `5.6.0-dev.202012021910.3a763f4`

**Comment by aparajita** [comment]
Thanks @liamdebeasi, I'll give it a shot and let you know what I find.

**Comment by aparajita** [comment]
> I am currently focusing on getting this working with webview resizing disabled.

Honestly, I don't know why anyone would **want** the webview to resize when the keyboard appears. It wreaks havoc on the layout.

**Comment by aparajita** [comment]
Unfortunately, @liamdebeasi, for me that dev build is not working on iOS. On the other hand, I discovered that if I use the current release build and set the keyboard resize mode to "body", the input fields are being scrolled into view perfectly.

**Comment by liamdebeasi** [comment]
> Unfortunately, @liamdebeasi, for me that dev build is not working on iOS.

Can you be more specific as to what is not working?

**Comment by aparajita** [comment]
Inputs are not scrolling into view. They move one pixel up I believe. This is with a different app than the input torture test app. I'll be pushing it to github today and you can try yourself.

**Comment by liamdebeasi** [comment]
Did you disable webview resizing on that app? I noticed in your other example your config does not set the resize mode properly: https://github.com/aparajita/ionic-input-problem/blob/main/capacitor.config.json#L12 (Should be `"resize": "none"` not `"resizeMode": "none"`)

**Comment by aparajita** [comment]
Yes, it was `"resize": "none"`.

**Comment by fpachecoibz** [comment]
I have the same problem. When I tap on an ion-input that would be below the keyboard, it takes 1.5s to focus:

![2021-04-08 11-35-05](https://user-images.githubusercontent.com/42489814/114012762-a6ef9900-985e-11eb-86f5-d7197ac10737.gif)

The strangest thing is that the ion-input is marked as touched before having the focus.

**Comment by fr3fou** [comment]
Any updates on this?

**Comment by liamdebeasi** [comment]
Hi everyone,

I went back to the drawing board on how best to fix this. Simply re-enabling the scroll padding utility would have caused adverse effects in situations where the padding is not needed. Additionally, there were race conditions where the padding would get added too late or removed entirely.

-----

I have a new dev build below which will fix the scroll padding issue on inputs at the bottom of a container:

```
6.2.4-dev.11661802454.1f8ad904
```

A few things to note:

1. **Despite the dev build saying "6.2.4" this is going to be an Ionic 7 fix.**

As a result, this dev build is subject to the [Ionic 7 Breaking Changes](https://github.com/ionic-team/ionic-framework/blob/feature-7.0/BREAKING.md#version-7x). Note: Ionic 7 is still in development.

**2. This patch only focuses on ensuring that the `--keyboard-offset` variable is set correctly**.

This variable adds `padding-bottom` to the scrollable area of your app's content to allow any inputs to be scrolled above the keyboard. Any other issue related to forms or scrolling inputs into view should be discussed [in new issues](https://github.com/ionic-team/ionic-framework/issues/new/choose).

**3. Scroll padding will only be added when your webview resize mode is set to `"none"` or if it is unknown.**
This means different things based on where your app is running:

- **Mobile Safari/PWA:** The keyboard always overlays the webview, so scroll padding will always be enabled.
- **Cordova:** We do not have support for detecting the webview resize mode in Cordova apps. As a result, scroll padding will always be enabled even if webview resizing is enabled too.
- **Capacitor:** Developers who are using `@capacitor/keyboard@4` or newer will only have scroll padding enabled if their [resize mode](https://capacitorjs.com/docs/apis/keyboard#keyboardresize) is set to `"none"`. This avoids the problem where blank space was being added on `ion-content` when it was not needed. Older versions of this plugin do not provide us with the necessary information, so scroll padding will be enabled regardless of resize mode on older versions.

------

Here is a before and after sample using code provided by @aparajita:

| old behavior | new behavior |
| ------------ | ------------- |
| <video src="https://user-images.githubusercontent.com/2721089/187297597-1b797e4b-8887-46fa-a1c3-70a7b50a68e9.mp4"></video> | <video src="https://user-images.githubusercontent.com/2721089/187297639-af8931f4-ab49-4a85-8fc6-b9285cbdc7b8.mp4"></video> |

Let me know if you run into any issues with the dev build. We also plan on having a beta version of Ionic 7 to get this code in front of more developers prior to release. Thanks!

**Comment by liamdebeasi** [comment]
Thanks for the issue. This has been resolved via https://github.com/ionic-team/ionic-framework/pull/25849, and a fix will be available in Ionic 7.0.

Please feel free to continue testing the dev build, and let me know if you run into any issues. There will be a public beta period for Ionic 7 for more testing as well.
